src/OSPSuite.Core/Domain/Builder/SimulationBuilder.cs

```diff
@@ -203,38 +285,131 @@ private void addMolecules(IMoleculeDependentBuilder builder, MoleculeList molecu
       internal IReadOnlyCollection<ParameterValue> ParameterValues => _parameterValues;
       internal IReadOnlyCollection<InitialCondition> InitialConditions => _initialConditions;
 
+      private IReadOnlyList<(TBuildingBlock buildingBlock, MergeBehavior mergeBehavior)> buildingBlockAndMergeBehaviors<TBuildingBlock>(Func<Module, TBuildingBlock> propAccess) where TBuildingBlock : class, IBuildingBlock =>
+         _simulationConfiguration.ModuleConfigurations
+            .Where(x => propAccess(x.Module) != null)
+            .Select(x => (propAccess(x.Module), x.MergeBehavior))
+            .ToList();
+
       public virtual IReadOnlyCollection<SimulationEntitySource> EntitySources => _entitySources;
 
-      internal MoleculeList MoleculeListFor(IMoleculeDependentBuilder builder) => _moleculeListCache[builder];
+      internal MoleculeList MoleculeListFor(IMoleculeDependentBuilder builder) => builder.MoleculeList;
 
       internal MoleculeBuilder MoleculeByName(string name) => _molecules[name];
 
-      internal class BuilderSource
+      internal class BuilderSource : BuilderSource<IEntity>
       {
-         public IEntity Builder { get; }
+         public BuilderSource(IEntity builder, IBuildingBlock buildingBlock) : base(builder, buildingBlock)
+         {
+         }
+      }
+
+      internal BuilderSource BuilderSourceFor(IEntity sourceEntity) => _builderSources[sourceEntity.Id];
+
+      internal class BuilderSource<T> where T : IEntity
+      {
+         public T Builder { get; }
          public IBuildingBlock BuildingBlock { get; }
 
-         public BuilderSource(IEntity builder, IBuildingBlock buildingBlock)
+         public BuilderSource(T builder, IBuildingBlock buildingBlock)
          {
             Builder = builder;
             BuildingBlock = buildingBlock;
          }
 
-         public override string ToString()
+         public void Deconstruct(out T builder, out IBuildingBlock buildingBlock)
          {
-            return $"{Builder.EntityPath()} - {BuildingBlock.Name}";
+            builder = Builder;
+            buildingBlock = BuildingBlock;
          }
       }
 
-      internal BuilderSource BuilderSourceFor(IEntity sourceEntity) => _builderSources[sourceEntity.Id];
+      internal class BuilderMergeInfo<T> where T : IBuilder
+      {
+         public BuilderSource<T> BaseBuilder { get; }
+         public IReadOnlyList<BuilderSource<T>> BuildersToMerge { get; }
```

Comment:
Or maybe 'BuilderMergeInfo` is the right name, but this should be `BuildersToExtendSource`.